Jenkinsfile

```diff
@@ -66,8 +66,41 @@ for(j = 0; j < jdks.size(); j++) {
                         if (! fileExists('test/target/surefire-reports/TEST-jenkins.Junit4TestsRanTest.xml') ) {
                             error 'junit 4 tests are no longer being run for the test package'
                         } // cli has been migrated to junit 5
+                        if (failFast && currentBuild.result == 'UNSTABLE') {
+                            error 'There were test failures; halting early'
+                        }
                     }
                     if (buildType == 'Linux' && jdk == jdks[0]) {
+                        def folders = env.JOB_NAME.split('/')
+                        if (folders.length > 1) {
+                            discoverGitReferenceBuild(scm: folders[1])
+                        }
+
+                        echo "Recording static analysis results for '${buildType}'"
+                        recordIssues enabledForFailure: true,
+                                tool: mavenConsole(),
+                                skipBlames: true,
+                                trendChartType: 'TOOLS_ONLY'
+                        recordIssues enabledForFailure: true,
+                                tools: [java(), javaDoc()],
+                                filters: [excludeFile('.*Assert.java')],
+                                sourceCodeEncoding: 'UTF-8',
+                                skipBlames: true,
+                                trendChartType: 'TOOLS_ONLY'
+                        recordIssues([tool: spotBugs(pattern: '**/target/spotbugsXml.xml'),
+                                                 sourceCodeEncoding: 'UTF-8',
+                                                 skipBlames: true,
+                                                 trendChartType: 'TOOLS_ONLY',
+                                                 qualityGates: [[threshold: 1, type: 'NEW', unstable: true]]])
```

Comment:
With that we could also enable spotbugs on medium and make it a lot harder that new issues get introduced.

Reply:
We could but how do you reproduce locally is it a matter of enabling medium and searching?

Not a huge fan of the approach but probably better to do so that new issues don’t creep in till we change the level

Reply:
> Not a huge fan of the approach

I think you're seeing why I didn't go that far in this PR. :-) Yes, that is a more controversial change. I'm not a fan of using the CI system as a ratchet either: I'd rather use the build system for that. But I don't feel strongly either way. In any case that shouldn't matter for this PR, which preserves the status quo.